### PR TITLE
Operator Feature for Dedicated Networking

### PR DESCRIPTION
**Pull request by joshuajerin**

### CODE REVIEW

File: tembo-operator/src/dedicated_networking.rs [hunk 1]
```diff
@@ -0,0 +1,568 @@
...
+/// - `cidr`: The CIDR block to allow traffic from.
+async fn reconcile_dedicated_networking_network_policies(
+    cdb: CoreDB,
+    client: Client,
+    namespace: &str,
+) -> Result<(), OperatorError> {
+    let cdb_name = cdb.name_any();
+    let np_api: Api<NetworkPolicy> = Api::namespaced(client, namespace);
+    let cidr = match cdb.spec.ip_allow_list.clone() {
+        None => "0.0.0.0/0".to_string(),
```
[the remark this answers is not in the document]
Comment (sjmiller609) [reply tembo-operator/src/dedicated_networking.rs:241]: this means allowing ingress networking from the cloud load balancer into this workload.

File: tembo-operator/src/dedicated_networking.rs [hunk 2]
```diff
@@ -0,0 +1,568 @@
...
+        "internet-facing"
+    } else {
+        "internal"
+    };
+    let lb_internal = if is_public { "false" } else { "true" };
+
+    info!(
+        "Applying Service: {} in namespace: {} with type: {} and scheme: {}",
+        service_name, namespace, service_type, lb_scheme
+    );
```
Comment (sjmiller609) [inline tembo-operator/src/dedicated_networking.rs:395]: debug

File: tembo-operator/src/dedicated_networking.rs [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+
+    let mut annotations = serde_json::Map::new();
+    annotations.insert(
+        "external-dns.alpha.kubernetes.io/hostname".to_string(),
+        serde_json::Value::String(format!("{}.{}", service_name, "example.com")),
```
Comment (sjmiller609) [inline tembo-operator/src/dedicated_networking.rs:400]: namespace, not service_name and basedomain, not example.com

File: tembo-operator/src/dedicated_networking.rs [hunk 4]
```diff
@@ -0,0 +1,568 @@
...
+    );
+
+    if service_type == "LoadBalancer" {
```
Comment (sjmiller609) [inline tembo-operator/src/dedicated_networking.rs:403]: Just add all annotations regardless of if it's type LoadBalancer or not

File: tembo-operator/src/dedicated_networking.rs [hunk 5]
```diff
@@ -0,0 +1,568 @@
...
+            "cnpg.io/cluster": cdb_name,
+            "role": if is_standby { "replica" } else { "primary" }
+        }),
+    );
+    service_spec.insert("sessionAffinity".to_string(), json!("None"));
+    service_spec.insert("type".to_string(), json!(service_type));
+
+    if service_type == "LoadBalancer" {
+        let load_balancer_source_ranges = match cdb.spec.ip_allow_list.clone() {
+            None => vec!["0.0.0.0/0".to_string()],
```
Comment (sjmiller609) [inline tembo-operator/src/dedicated_networking.rs:466]: instead of defaulting to all IPs, I think we don't add loadBalancerSourceRanges to the spec

Comment (sjmiller609) [reply tembo-operator/src/dedicated_networking.rs:466]: then when we QA, we need to check it works with or without IP allow list set

File: tembo-operator/src/dedicated_networking.rs [hunk 6]
```diff
@@ -0,0 +1,568 @@
...
+            "namespace": namespace,
+            "annotations": annotations,
+            "ownerReferences": [owner_reference],
+            "labels": labels
+        },
+        "spec": service_spec
+    });
+
+    let svc_api: Api<Service> = Api::namespaced(client, namespace);
+    let patch_params = PatchParams::apply("conductor").force();
```
Comment (sjmiller609) [inline tembo-operator/src/dedicated_networking.rs:489]: this should not say conductor

File: tembo-operator/src/dedicated_networking.rs [hunk 7]
```diff
@@ -0,0 +1,568 @@
...
+    } else {
+        format!("{}-dedicated", cdb_name)
+    };
+
+    let svc_api: Api<Service> = Api::namespaced(client, namespace);
+
+    info!(
+        "Checking if service: {} exists in namespace: {} for deletion",
+        service_name, namespace
+    );
```
Comment (sjmiller609) [inline tembo-operator/src/dedicated_networking.rs:537]: debug

File: tembo-operator/src/dedicated_networking.rs [hunk 8]
```diff
@@ -0,0 +1,568 @@
...
+
+    if svc_api.get(&service_name).await.is_ok() {
+        info!(
+            "Service: {} exists in namespace: {}. Proceeding with deletion.",
+            service_name, namespace
+        );
```
Comment (sjmiller609) [inline tembo-operator/src/dedicated_networking.rs:543]: this is good at info level

File: tembo-operator/src/ingress.rs [hunk 9]
```diff
@@ -250,6 +250,12 @@ pub async fn reconcile_postgres_ing_route_tcp(
     middleware_names: Vec<String>,
     delete: bool,
 ) -> Result<(), OperatorError> {
+    // Check if the basedomain is set
+    if basedomain.is_empty() {
+        debug!("Base domain is not set. Skipping IngressRouteTCP reconciliation.");
+        return Ok(());
+    }
+
```
Comment (sjmiller609) [inline tembo-operator/src/ingress.rs:258]: this seems like an unrelated change

File: tembo-operator/yaml/sample-dedicated-networking.yaml [hunk 10]
```diff
@@ -0,0 +1,11 @@
+apiVersion: coredb.io/v1alpha1
+kind: CoreDB
+metadata:
+  name: sample-dedicated
```
Comment (sjmiller609) [inline tembo-operator/yaml/sample-dedicated-networking.yaml:4]: here is the place to set as 30 character name

File: tembo-operator/tests/integration_tests.rs [hunk 11]
```diff
@@ -1837,14 +1900,13 @@ mod test {
...
-        // Initialize the Kubernetes client
+    async fn test_networking() {
         let client = kube_client().await;
+        let state = State::default();
 
-        // Configurations
         let mut rng = rand::thread_rng();
         let suffix = rng.gen_range(0..100000);
-        let name = &format!("test-ingress-route-tcp-{}", suffix.clone());
+        let name = &format!("test-networking-{}", suffix.clone());
```
Comment (sjmiller609) [inline tembo-operator/tests/integration_tests.rs:1909]: or maybe here is the place to make sure it's 30 characters

File: tembo-operator/tests/integration_tests.rs [hunk 12]
```diff
@@ -2037,12 +2080,121 @@ mod test {
...
+                    "public": true,
+                    "serviceType": "LoadBalancer"
+                }
+            }
+        });
+        let params = PatchParams::apply("functional-test-dedicated-networking");
+        let patch = Patch::Apply(&coredb_json);
+        let _coredb_resource = coredbs.patch(name, &params, &patch).await.unwrap();
+
+        tokio::time::sleep(Duration::from_secs(10)).await;
```
Comment (sjmiller609) [inline tembo-operator/tests/integration_tests.rs:2108]: this looks like a sleep in order to wait for something, but instead it should have a wait retry loop to prevent test error if cluster is slow or busy

File: tembo-operator/src/dedicated_networking.rs [hunk 13]
```diff
@@ -0,0 +1,568 @@
...
+            info!(
+                "Reconciling IP allow list middleware for CoreDB instance: {}",
+                cdb.name_any()
+            );
+            let middleware_name = reconcile_ip_allowlist_middleware(cdb, ctx.clone())
+                .await
+                .map_err(|e| {
+                    error!("Failed to reconcile IP allowlist middleware: {:?}", e);
+                    e
+                })?;
```
[the remark this answers is not in the document]
Comment (sjmiller609) [reply tembo-operator/src/dedicated_networking.rs:77]: I think this comment still needs addressed

File: tembo-operator/src/dedicated_networking.rs [hunk 14]
```diff
@@ -0,0 +1,575 @@
...
+                port.clone(),
+                vec![middleware_name.clone()],
+                false,
+                false,
+            )
+            .await
+            .map_err(|e| {
+                error!("Failed to reconcile primary ingress route: {:?}", e);
+                e
+            })?;
```
Comment (sjmiller609) [inline tembo-operator/src/dedicated_networking.rs:113]: I think there should be no ingress route TCP created for this

File: tembo-operator/src/dedicated_networking.rs [hunk 15]
```diff
@@ -0,0 +1,575 @@
...
+                    &format!("{}-dedicated-ro", cdb.name_any()),
+                    port.clone(),
+                    vec![middleware_name],
+                    false,
+                    true,
+                )
+                .await
+                .map_err(|e| {
+                    error!("Failed to reconcile standby ingress route: {:?}", e);
+                    e
```
Comment (sjmiller609) [inline tembo-operator/src/dedicated_networking.rs:149]: also remove the ingress route tcp here, should not be added

File: tembo-operator/src/dedicated_networking.rs [hunk 16]
```diff
@@ -0,0 +1,575 @@
...
+                })?;
+            }
```
Comment (sjmiller609) [inline tembo-operator/src/dedicated_networking.rs:151]: If standby is disabled, it should delete the standby service. For example - enable HA, enable dedicated networking, disable HA. Should end up with 1 dedicated service only

File: tembo-operator/src/dedicated_networking.rs [hunk 17]
```diff
@@ -0,0 +1,575 @@
...
+                ctx,
+                &ns,
+                basedomain,
+                "dedicated-ro",
+                &format!("{}-dedicated-ro", cdb.name_any()),
+                port,
+                vec![],
+                true,
+                true,
+            )
```
Comment (sjmiller609) [inline tembo-operator/src/dedicated_networking.rs:201]: remove handling of ingress route tcp from this PR

File: tembo-operator/src/dedicated_networking.rs [hunk 18]
```diff
@@ -0,0 +1,575 @@
...
+        .unwrap_or_else(|_| "10.0.0.0/8".to_string())
+        .split(',')
+        .map(|s| s.trim().to_string())
+        .collect::<Vec<String>>();
+
+    let policy_name = format!("{}-allow-nlb", cdb_name);
+    info!(
+        "Applying network policy: {} in namespace: {} to allow traffic from CIDRs: {:?}",
+        policy_name, namespace, cidr_list
+    );
```
Comment (sjmiller609) [inline tembo-operator/src/dedicated_networking.rs:251]: debug

File: tembo-operator/src/dedicated_networking.rs [hunk 19]
```diff
@@ -0,0 +1,575 @@
...
+                "Failed to apply network policy: {} in namespace: {}. Error: {:?}",
+                policy_name, namespace, e
+            );
+            OperatorError::NetworkPolicyError(format!("Failed to apply network policy: {:?}", e))
+        })?;
+
+    info!(
+        "Successfully applied network policy: {} in namespace: {}",
+        policy_name, namespace
+    );
```
Comment (sjmiller609) [inline tembo-operator/src/dedicated_networking.rs:305]: debug

File: tembo-operator/src/dedicated_networking.rs [hunk 20]
```diff
@@ -0,0 +1,575 @@
...
+/// - `namespace`: The namespace in which to apply the ingress routes.
+/// - `basedomain`: The base domain for the ingress routes.
+/// - `ingress_name_prefix`: The prefix for the ingress resource names.
+/// - `service_name`: The name of the service associated with the ingress route.
+/// - `port`: The port for the service.
+/// - `middleware_names`: A list of middleware names to associate with the ingress route.
+/// - `delete`: Whether to delete the ingress route.
+/// - `is_standby`: Whether the ingress route is for a standby service.
+#[allow(clippy::too_many_arguments)]
+async fn reconcile_dedicated_networking_ing_route_tcp(
```
Comment (sjmiller609) [inline tembo-operator/src/dedicated_networking.rs:326]: delete function

File: tembo-operator/src/dedicated_networking.rs [hunk 21]
```diff
@@ -0,0 +1,575 @@
...
+
+    debug!(
+        "Applying Service: {} in namespace: {} with type: {} and scheme: {}",
+        service_name, namespace, service_type, lb_scheme
+    );
+
+    let mut annotations = serde_json::Map::new();
+    annotations.insert(
+        "external-dns.alpha.kubernetes.io/hostname".to_string(),
+        serde_json::Value::String(format!("{}", namespace)),
```
Comment (sjmiller609) [inline tembo-operator/src/dedicated_networking.rs:407]: this is incorrect, should have namespace.basedomain

File: tembo-operator/tests/integration_tests.rs [hunk 22]
```diff
@@ -2037,12 +2080,119 @@ mod test {
...
+            service
+                .metadata
+                .labels
+                .as_ref()
+                .expect("Labels should be present")
+                .get("public")
+                .expect("Public label should be present"),
+            "true"
+        );
+
```
Comment (sjmiller609) [inline tembo-operator/tests/integration_tests.rs:2157]: please add an assertion of what the hostname should look like on the DNS annotation

File: tembo-operator/tests/integration_tests.rs [hunk 23]
```diff
@@ -1837,14 +1900,13 @@ mod test {
...
-        // Initialize the Kubernetes client
+    async fn test_networking() {
         let client = kube_client().await;
+        let state = State::default();
 
-        // Configurations
         let mut rng = rand::thread_rng();
         let suffix = rng.gen_range(0..100000);
-        let name = &format!("test-ingress-route-tcp-{}", suffix.clone());
+        let name = &format!("example-dedicated-networking-{}", suffix.clone());
```
Comment (sjmiller609) [inline tembo-operator/tests/integration_tests.rs:1909]: Please change so this name variable is exactly 30 characters long after formatting.

**Review by sjmiller609 — CHANGES_REQUESTED** [review]

File: tembo-operator/src/dedicated_networking.rs [hunk 24]
```diff
@@ -0,0 +1,443 @@
...
+                    error!("Failed to delete standby service: {:?}", e);
+                    e
+                })?;
+        }
+    } else {
+        debug!(
+            "Dedicated networking is not configured for CoreDB instance: {}",
+            cdb.name_any()
+        );
+    }
```
Comment (sjmiller609) [inline tembo-operator/src/dedicated_networking.rs:136]: In the case of dedicated networking not enabled, it should still clean up resources. For example, enable dedicated networking, then delete the configuration from coredb, it should clean up the resources

File: tembo-operator/src/dedicated_networking.rs (hunk 2 above)
Comment (sjmiller609) [reply tembo-operator/src/dedicated_networking.rs:270]: this should be debug

File: tembo-operator/src/dedicated_networking.rs (hunk 19 above)
Comment (sjmiller609) [reply tembo-operator/src/dedicated_networking.rs:228]: this should be debug

File: tembo-operator/src/dedicated_networking.rs (hunk 18 above)
Comment (sjmiller609) [reply tembo-operator/src/dedicated_networking.rs:174]: this should be debug

File: tembo-operator/src/dedicated_networking.rs [hunk 25]
```diff
@@ -0,0 +1,443 @@
...
+                "Failed to apply service: {} in namespace: {}. Error: {}",
+                service_name, namespace, e
+            );
+            OperatorError::ServiceError(format!("Failed to apply service: {:?}", e))
+        })?;
+
+    info!(
+        "Successfully applied service: {} in namespace: {}",
+        service_name, namespace
+    );
```
Comment (sjmiller609) [inline tembo-operator/src/dedicated_networking.rs:381]: this should be debug

File: tembo-operator/src/dedicated_networking.rs (hunk 7 above)
Comment (sjmiller609) [reply tembo-operator/src/dedicated_networking.rs:412]: this should be debug

File: tembo-operator/src/dedicated_networking.rs [hunk 26]
```diff
@@ -0,0 +1,443 @@
...
+    info!(
+        "Checking if service: {} exists in namespace: {} for deletion",
+        service_name, namespace
+    );
+
+    if svc_api.get(&service_name).await.is_ok() {
+        debug!(
+            "Service: {} exists in namespace: {}. Proceeding with deletion.",
+            service_name, namespace
+        );
```
Comment (sjmiller609) [inline tembo-operator/src/dedicated_networking.rs:418]: this should be info

File: tembo-operator/tests/integration_tests.rs [hunk 27]
```diff
@@ -1837,14 +1900,13 @@ mod test {
...
+    async fn test_networking() {
         let client = kube_client().await;
+        let state = State::default();
 
-        // Configurations
         let mut rng = rand::thread_rng();
-        let suffix = rng.gen_range(0..100000);
-        let name = &format!("test-ingress-route-tcp-{}", suffix.clone());
+        let suffix = rng.gen_range(0..10000);
+        let name = &format!("test-dedicated-networking-{}", suffix.clone());
```
Comment (sjmiller609) [inline tembo-operator/tests/integration_tests.rs:1909]: name variable should be exactly 30 characters, after the formatting.

**Review by sjmiller609 — CHANGES_REQUESTED** [review]

File: tembo-operator/src/dedicated_networking.rs [hunk 28]
```diff
@@ -0,0 +1,568 @@
...
+    is_standby: bool,
+    service_type: &str,
+) -> Result<(), OperatorError> {
+    let cdb_name = &cdb.name_any();
+    let owner_reference = cdb.controller_owner_ref(&()).unwrap();
+    let service_name = if is_standby {
+        format!("{}-dedicated-ro", cdb_name)
+    } else {
+        format!("{}-dedicated", cdb_name)
+    };
```
[the remark this answers is not in the document]
Comment (joshuajerin) [reply tembo-operator/src/dedicated_networking.rs:272]: Added the length of the name in the integration test to test it out

File: tembo-operator/yaml/sample-dedicated-networking.yaml (hunk 10 above)
Comment (joshuajerin) [reply tembo-operator/yaml/sample-dedicated-networking.yaml:4]: Added the 30 character name in the integration tests. Skipping this

File: tembo-operator/tests/integration_tests.rs (hunk 27 above)
Comment (joshuajerin) [reply tembo-operator/tests/integration_tests.rs]: Changed the range from 0 to 1000

**Review by sjmiller609 — APPROVED** [review]
